Introduce assertEqualIgnoreType to prevent type checking of assertEqual

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38195 Replace instance of  assertEqualIgnoreType
* #38152 Tests: Remove exact_dtype from test classes
* #38124 Remove asserEqualIgnoreType from test_complex
* #38123 Fix tests for non MKL enviroments
* #38103 assertEqual now requires matching dtypes
* #38102 Replacing assertEqual with assertEqualIgnoreType wherever types missmatch
* **#34154 Introduce assertEqualIgnoreType to prevent type checking of assertEqual**

Will be temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

Differential Revision: [D20251131](https://our.internmc.facebook.com/intern/diff/D20251131)

VitalyFedyunin committed:
AssertEqual now checks tensors dtype

Temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "AssertEqual now checks tensors dtype"


Temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "AssertEqual now checks tensors dtype"


Temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

Differential Revision: [D20251131](https://our.internmc.facebook.com/intern/diff/D20251131)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Introduce assertEqualIgnoreType to prevent type checking of assertEqual"


Will be temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

Differential Revision: [D20251131](https://our.internmc.facebook.com/intern/diff/D20251131)

[ghstack-poisoned]

mruberry commented:
Is the idea that this will be easier to search for instead of just setting exact_dtype=False? That also seems kinda easy to search for?

VitalyFedyunin commented:
Sometimes it is valid to setup exact_dtype, also idea is to put the only with todo lines (see next two PRs in the stack) and follow up with step by step removal (see #38124).

mruberry reviewed (approved):
Cool!

VitalyFedyunin committed:
Rebase to master on "Introduce assertEqualIgnoreType to prevent type checking of assertEqual"


Will be temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

Differential Revision: [D20251131](https://our.internmc.facebook.com/intern/diff/D20251131)

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase to master on "Introduce assertEqualIgnoreType to prevent type checking of assertEqual"


Will be temporary replacing with `assertEqualIgnoreType` all cases when `AssertEqual` fails.

Differential Revision: [D20251131](https://our.internmc.facebook.com/intern/diff/D20251131)

[ghstack-poisoned]
